src/lib.rs

```diff
@@ -1,287 +1,26 @@
-}
-
-/// Same as [DeviceMmio] but the methods are invoked with a mutable self borrow.
-/// # Example
-/// ```
-/// # use vm_device::{MutDeviceMmio, bus::{MmioAddress, MmioAddressOffset}};
-/// struct DummyDevice {
-///     config: u32,
-/// }
-///
-/// impl MutDeviceMmio for DummyDevice {
-///     fn mmio_read(&mut self, _base: MmioAddress, _offset: MmioAddressOffset, data: &mut [u8]) {
-///         if data.len() > 4 {
-///             return;
-///         }
-///         for (idx, iter) in data.iter_mut().enumerate() {
-///             *iter = (self.config >> (idx * 8) & 0xff) as u8;
-///         }
-///     }
-///
-///     fn mmio_write(&mut self, _base: MmioAddress, _offset: MmioAddressOffset, data: &[u8]) {
-///         self.config = u32::from(data[0]) & 0xff;
-///     }
-/// }
-/// ```
-pub trait MutDeviceMmio {
-    /// Handle a read operation on the device.
-    ///
-    /// # Arguments
-    ///
-    /// * `base`:   base address on a MMIO bus
-    /// * `offset`: base address' offset
-    /// * `data`:   a buffer provided by the caller to store the read data
-    fn mmio_read(&mut self, base: MmioAddress, offset: MmioAddressOffset, data: &mut [u8]);
-
-    /// Handle a write operation to the device.
-    ///
-    /// # Arguments
-    ///
-    /// * `base`:   base address on a MMIO bus
-    /// * `offset`: base address' offset
-    /// * `data`:   a buffer provided by the caller holding the data to write
-    fn mmio_write(&mut self, base: MmioAddress, offset: MmioAddressOffset, data: &[u8]);
-}
+pub use virtio_mmio::VirtioMmioOffset;
 
-// Blanket implementations for Arc<T>.
-
-impl<T: DeviceMmio + ?Sized> DeviceMmio for Arc<T> {
-    fn mmio_read(&self, base: MmioAddress, offset: MmioAddressOffset, data: &mut [u8]) {
-        self.deref().mmio_read(base, offset, data);
-    }
-
-    fn mmio_write(&self, base: MmioAddress, offset: MmioAddressOffset, data: &[u8]) {
-        self.deref().mmio_write(base, offset, data);
-    }
-}
-
-impl<T: DevicePio + ?Sized> DevicePio for Arc<T> {
-    fn pio_read(&self, base: PioAddress, offset: PioAddressOffset, data: &mut [u8]) {
-        self.deref().pio_read(base, offset, data);
-    }
-
-    fn pio_write(&self, base: PioAddress, offset: PioAddressOffset, data: &[u8]) {
-        self.deref().pio_write(base, offset, data);
-    }
-}
-
-// Blanket implementations for Mutex<T>.
-
-impl<T: MutDeviceMmio + ?Sized> DeviceMmio for Mutex<T> {
-    fn mmio_read(&self, base: MmioAddress, offset: MmioAddressOffset, data: &mut [u8]) {
-        self.lock().unwrap().mmio_read(base, offset, data)
-    }
-
-    fn mmio_write(&self, base: MmioAddress, offset: MmioAddressOffset, data: &[u8]) {
-        self.lock().unwrap().mmio_write(base, offset, data)
-    }
+pub trait VirtioMmioDevice {
+    fn virtio_mmio_read(&self, base: GuestAddress, offset: VirtioMmioOffset, data: &mut [u8]);
+    fn virtio_mmio_write(&self, base: GuestAddress, offset: VirtioMmioOffset, data: &[u8]);
```

Comment:
This commit is huge and highly intrusive. Since it seems to be turning the `vm_device` crate into a virtio specific one, I have a couple of fundamental questions:

- How is a virtio MMIO device different than a `DeviceMmio` one?
- You're renaming a bunch of structures, deleting a bunch of modules, to essentially accommodate for  virtio MMIO. But is there something missing from the `vm_device` crate that would prevent you to use it as is?

Reply:
- A virtio MMIO device is basically a `DeviceMmio`, with some slightly tweaked QoL (such as having the offset as an Enum corresponding to the virtio layout, or using `GuestAddress` , the same as in lumper). They are essentially the same.
- Nothing is missing from `vm_device`. However, this fork is easier to work with, since it corresponds to our use case. 

I will try giving another shot at using the upstream crate.